File: Codes/2786-visit-array-positions-to-maximize-score.go
/*
 * @Author: LetMeFly
 * @Date: 2024-06-14 19:07:00
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-06-14 19:13:41
 */
package main

func max(a int64, b int64) int64 {
    if a > b {
        return a
    }
    return b
}

func maxScore(nums []int, x int) int64 {
    odd, even := int64(0), int64(0)
    if nums[0] % 2 == 0 {
        odd = -int64(x)
    } else {
        even = -int64(x)
    }
    for _, t := range nums {
        if t % 2 != 0 {
            odd = max(odd + int64(t), even + int64(t) - int64(x))
        } else {
            even =  max(odd + int64(t) - int64(x), even + int64(t))
        }
    }
    return max(odd, even)
}

File: Codes/2786-visit-array-positions-to-maximize-score.java
/*
 * @Author: LetMeFly
 * @Date: 2024-06-14 19:04:20
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-06-14 19:06:28
 */
class Solution {
    public long maxScore(int[] nums, int x) {
        long odd = nums[0] % 2 != 0 ? 0 : -x, even = nums[0] % 2 != 0 ? -x : 0;
        for (int t : nums) {
            if (t % 2 != 0) {
                odd = Math.max(odd + t, even + t - x);
            }
            else {
                even = Math.max(odd + t - x, even + t);
            }
        }
        return Math.max(odd, even);
    }
}